keps/sig-api-machinery/4460-per-request-deadline/README.md

```diff
@@ -0,0 +1,916 @@
+	RequestURI     string
+	RequestAuditID string
+}
+
+type tracker struct{
+	inflight map[*http.Request]*RequestData
+}
+
+// Inside the WithRequestDeadline filetr
+
+func(w http.ResponseWriter, req *http.Request) {
+...
+	if utilfeature.DefaultFeatureGate.Enabled(features.PerHandlerReadWriteTimeout) {
+		data := &RequestData{StartedAt: startedAt, DeadlineAt: deadlineAt, RequestURI: "", AuditID: ""}
+		tracker.Enter(req, data)
+		defer tracker.Leave(req)	
+	}
+	handler.ServeHTTP(w, req)
+...
+```
+
+The `PostStartHook` goroutine will wake up every `5m`, and perform the following operation:
+
+- Iterate through the map entries 
+- If an entry is resident for more than `5m` after the specified `DeadlineAt` then
+  it is reported in the metrics/logs
+- If an entry is resident for more than `10m`, we remove it from the hashmap so the
+  list does not grow infinitely.
+  
+
+The memory cost of the above hashmap is to the `O(number of non long-running requests in flight)`.
+  
+One tradeoff of this global hashmap is the locking contention present due to
+in-flight requests trying to access the synchronized map concurrently. But it
+avoids creating a new goroutine for each request.
+
+
+
+##### Impact on the Client
+
+It's important to understand how the the client is impacted with this feature enabled; 
+upon timeout of a request, what the client sees depends on the following factors:
+
+- a) Protocol of the request, HTTP/1x or HTTP/2.0.
+- b) Whether the client specifies a timeout in the `timeout` parameter of the request.
+- c) Whether the request handler writes to the `ResponseWriter` object by invoking
+  `Write`, before the deadline exceeds.
+- d) Whether the request handler invokes `Flush` on the `ResponseWriter` object, 
+  preceded  by a `Write`, before the deadline exceeds.
+
+If enabling this feature introduces a change in what the client observes, then it should be 
+noted, and must be documented with a test. Let's outline the current behavior today, with
+the `timeout` filter enabled:
+
+When a request times out, the client today receives a `504 GatewayTimeout` only if the
+request handler does not write to the underlying `ResponseWriter` object before
+timeout occurs.
+
+Moreover, if the client sets a timeout value in the request URI `{host/path}?timeout=10s` 
+it will yield the same deadline for the request on both the client and the server. 
+This is the case with `client-go` today. In this scenario, if a request times out, 
+the client is most likely to receive a `context deadline exceeded` error due to
+the fact that the client will see the `Context` of the client request be expired
+well before the `504 GatewayTimeout` response from the server makes the round-trip
+to the client.
+
+Also, the client can not control when or whether the request handler writes to
+the `ResponseWriter` object. So, today a client can not deterministically catch
+the `504 GatewayTimeout` error from the server when a request times out.
```

Comment:
So what happens instead in those other cases?

Reply:
I have added the different scenarios, and what both the server and client sees with http/1x or http/2.0 the feature enabled/disabled